pkg/template/processor_test.go

```diff
@@ -0,0 +1,807 @@
+			require.NoError(t, err)
+			assert.Equal(t, "all-services", service.Spec.Selector["run"])
+		})
+
+		t.Run("when using forceUpdate=false, it should NOT update when using same object", func(t *testing.T) {
+			// given
+			cli := test.NewFakeClient(t)
+			processor := template.NewProcessor(cli, s)
+			_, err := processor.ApplySingle(defaultObject.DeepCopyObject(), true, nil)
+			require.NoError(t, err)
+
+			// when
+			createdOrChanged, err := processor.ApplySingle(defaultObject, false, nil)
+
+			// then
+			require.NoError(t, err)
+			assert.False(t, createdOrChanged)
+		})
+
+		t.Run("when object is missing, it should create it no matter what is set as forceUpdate", func(t *testing.T) {
+			// given
+			cli := test.NewFakeClient(t)
+			processor := template.NewProcessor(cli, s)
+			require.NoError(t, err)
+			deployment := &v1.Deployment{}
+
+			// when
+			createdOrChanged, err := processor.ApplySingle(modifiedObjs[0].Object, false, deployment)
+
+			// then
+			require.NoError(t, err)
+			assert.True(t, createdOrChanged)
+			service := &corev1.Service{}
+			err = cli.Get(context.TODO(), namespacedName, service)
+			require.NoError(t, err)
+			assert.Equal(t, "all-services", service.Spec.Selector["run"])
+			assert.NotEmpty(t, service.OwnerReferences)
+		})
+	})
+
+	t.Run("when using forceUpdate=false, it should update ConfigMap when data field is different", func(t *testing.T) {
+		// given
+		cli := test.NewFakeClient(t)
+		processor := template.NewProcessor(cli, s)
+		_, err := processor.ApplySingle(defaultObjs[1].Object, true, nil)
+		require.NoError(t, err)
+
+		// when
+		createdOrChanged, err := processor.ApplySingle(modifiedObjs[1].Object, false, nil)
+
+		// then
+		require.NoError(t, err)
+		assert.True(t, createdOrChanged)
+		configMap := &corev1.ConfigMap{}
+		namespacedName := types.NamespacedName{Namespace: "toolchain-host-operator", Name: "registration-service"}
+		err = cli.Get(context.TODO(), namespacedName, configMap)
+		require.NoError(t, err)
+		assert.Equal(t, "all-services", configMap.Data["service-selector"])
+	})
+}
+
+func addToScheme(t *testing.T) *runtime.Scheme {
+	s := scheme.Scheme
+	err := authv1.Install(s)
+	require.NoError(t, err)
+	err = apis.AddToScheme(s)
+	require.NoError(t, err)
+	return s
+}
+
+func getNameWithTimestamp(prefix string) string {
+	return fmt.Sprintf("%s-%d", prefix, time.Now().UnixNano())
+}
+
+func assertObject(t *testing.T, expectedObj expectedObj, actual runtime.RawExtension) {
+	// objJson, err := actual.Marshal()
+	// require.NoError(t, err, "failed to marshal json from unstructured object")
+	expected, err := newObject(expectedObj.template, expectedObj.username, expectedObj.commit)
+	require.NoError(t, err, "failed to create object from template")
+	assert.Equal(t, expected, actual.Object)
+}
+
+type expectedObj struct {
+	template string
+	username string
+	commit   string
+}
+
+func newObject(template, username, commit string) (*unstructured.Unstructured, error) {
```

Comment:
should you return a `runtime.Unstructured` here?

Reply:
that was already replaced here 657e55c  - you reviewed old code